File: clojure/39.clj
;; If p is the perimeter of a right angle triangle with integral length sides,
;; {a,b,c}, there are exactly three solutions for p = 120.
;;
;; {20,48,52}, {24,45,51}, {30,40,50}
;;
;; For which value of p ≤ 1000, is the number of solutions maximised?

(load "util")

(def pythagorean-triples-by-perimiter
  (letfn [(triples-with-perimiter [p acc]
            (let [lower-bound (inc (int (Math/sqrt (/ p 4))))
                  upper-bound (Math/sqrt (/ p 2))
                  candidates (filter #(divides? (* 2 %) p)
                                     (range lower-bound upper-bound))
                  new-triples (for [m candidates
                                    :let [n (- (/ p (* 2 m)) m)
                                          a (- (* m m) (* n n))
                                          b (* 2 m n)
                                          c (+ (* m m) (* n n))]]
                                [(min a b) (max a b) c])
                  old-triples (for [d (proper-divisors p)
                                    ts (nth acc d)]
                                (map #(* (quot p d) %) ts))]
              (distinct (concat new-triples old-triples))))
          (f [n acc]
            (let [ts (if (odd? n) () (triples-with-perimiter n acc))]
              (lazy-seq (cons ts (f (inc n) (conj acc ts))))))]
    (f 0 [])))

(defn euler-39 [n]
  (max-index (map count (take n pythagorean-triples-by-perimiter))))

(run (euler-39 1000))

File: haskell/39.hs
-- If p is the perimeter of a right angle triangle with integral length sides,
-- {a,b,c}, there are exactly three solutions for p = 120.
--
-- {20,48,52}, {24,45,51}, {30,40,50}
--
-- For which value of p ≤ 1000, is the number of solutions maximised?

import Util
import Data.List

triplesWithPerimiter p = distinct $ newTriples ++ oldTriples
  where
    lower      = succ . floor   . sqrt . (/ 4) . fromIntegral
    upper      = pred . ceiling . sqrt . (/ 2) . fromIntegral
    candidates = filter ((`divides` p) . (2 *)) [lower p .. upper p]
    newTriples = map (sort . euclidTr) candidates
    oldTriples = [map (* (quot p d)) ts | d <- properDivisors p,
                                          ts <- triplesWithPerimiter d]
    euclidTr m = let n = quot p (2 * m) - m in [m^2 - n^2, 2 * m * n, m^2 + n^2]

euler39 n = maxIndex $ map (length . triplesWithPerimiter) [0 .. pred n]

main = run $ euler39 1000
